lib/segment/src/index/field_index/map_index/mod.rs

```diff
@@ -362,6 +419,47 @@ where
     }
 }
 
+pub struct MapIndexMmapBuilder<N: MapIndexKey + ?Sized> {
+    path: PathBuf,
+    point_to_values: Vec<Vec<N::Owned>>,
+    values_to_points: HashMap<N::Owned, Vec<PointOffsetType>>,
+}
+
+impl<N: MapIndexKey + ?Sized> FieldIndexBuilderTrait for MapIndexMmapBuilder<N>
+where
+    MapIndex<N>: PayloadFieldIndex + ValueIndexer,
+    <MapIndex<N> as ValueIndexer>::ValueType: Into<N::Owned>,
+{
+    type FieldIndexType = MapIndex<N>;
+
+    fn init(&mut self) -> OperationResult<()> {
+        Ok(())
+    }
+
+    fn add_point(&mut self, id: PointOffsetType, payload: &[&Value]) -> OperationResult<()> {
+        let mut flatten_values: Vec<_> = vec![];
+        for value in payload.iter() {
+            let payload_values = <MapIndex<N> as ValueIndexer>::get_values(value);
+            flatten_values.extend(payload_values);
+        }
+        let flatten_values: Vec<N::Owned> = flatten_values.into_iter().map(Into::into).collect();
+        self.point_to_values.resize_with(id as usize + 1, Vec::new);
```

Comment:
This part makes building Mmap index very slow if there are many points to insert.
Simply passing the max point id to this function gave me a speedup from ~15-20 minutes to a few seconds.
I'd propose to pass a hint to this function with the largest point id and directly allocating the right size.
If you agree, I can make a PR!

Reply:
I agree, you can provide max id into the `init` method. You can also create your branch from this to avoid future merge conflicts
https://github.com/qdrant/qdrant/pull/4809

Reply:
Will do!